### DISCUSSION TITLE
Custom LFS protocol handlers

### DISCUSSION THREAD

**Discussion by axkibe** [git-lfs/git-lfs#5277] [Q&A] [answered]
Question: is there an easy way to plug in custom LFS handlers?

Rationale: On my LFS enabled (self coded) git server, I want to locally clone a repository sitting as bare repository in say file:///git/myrepo.git/. Except of course if I use a file:// link, LFS doesn't know how and where to get the LFS objects. Right now I worked around it, by using https://localhost/myrepo.git.. except this didn't work, since my SSL Certificate of course isn't for localhost, so I had to replace it with my hostname. And I don't like the idea there are issues waiting here for a local operation, when there is something with SSL (ceritifcate expiry and so on).

In case of cloning it could directly find the LFS objects on the filesystem gzipped at /lfs/[oid]:[length]. In case of pushing from said repository, however it should go through my server, since it is handling semaphoring, failed uploads and storing in a database what objects are used in which repository (just in case), perhaps I could code on both ends using a unix socket?

PS the rationale behind the rationale, I want to keep some git repositories in sync with overleaf.. and they don't support LFS, and converting back and forth didn't work for me at all. Thats why I clone both repositories (mine with LFS and the one from overleaf without) programmically and rsync between them after a git-receive-pack to couple them.

**Comment by bk2204** [comment, marked as answer]
Git LFS does indeed support cloning local files with `file:///` since around 2.10.  Note that the LFS files must actually be in the repo from which you're cloning under `.git/lfs` for this to work, and the destination repo must have the proper permissions (that is, `git -C /git/myrepo.git rev-parse HEAD` must succeed without problems).

There are also [custom transfer adapters](https://github.com/git-lfs/git-lfs/blob/main/docs/custom-transfers.md) which you can use.  If you want a kind that doesn't use the batch API, then you want the standalone variant, and our file protocol adapter is an example of this which you can look at.

**Comment by axkibe** [comment]
A thanks, custom transfer adapters is exactly what I was looking for.

(permissions are no issue, it is all done by one user on the file system.. which as a main reason behind the custom server that checks permissions  (similar to gitolite) since more complicated setups are a PITA to reflect in unix permissions.)
